### DISCUSSION TITLE
gre+ipsec tunnel, one router is behind NAT, TS_UNACCEPT

### DISCUSSION THREAD

**Discussion by Giveback-ru** [strongswan/strongswan#1915] [Q&A]
I have one VM behind a NAT in the public cloud, I'm trying to set up a GRE tunnel with ipsec to a remote party using strongswan. VM's NAT'ed public IP is 158.160.32.170, and the IP which the VM is aware of is 10.10.1.13 (for those who have read my previous question - yeah, the same VM, but the case is slightly different). Remote side is running the infamous RouterOS, known for it's lack of VTI support for a decade now. I've tried various local_ts/remote_ts combination but so far I'm only receving "traffic selectors unacceptable" either from a local or from a remote side:

````
Sep 19 14:47:14 web-ng charon-systemd[261069]: received packet: from 5.227.67.50[4500] to 10.10.1.13[4500] (400 bytes)
Sep 19 14:47:14 web-ng charon-systemd[261069]: parsed CREATE_CHILD_SA request 38 [ No SA TSi TSr N(USE_TRANSP) ]
Sep 19 14:47:14 web-ng charon-systemd[261069]: traffic selectors 158.160.32.170/32 === 5.227.67.50/32 unacceptable
Sep 19 14:47:14 web-ng charon-systemd[261069]: failed to establish CHILD_SA, keeping IKE_SA
Sep 19 14:47:14 web-ng charon-systemd[261069]: generating CREATE_CHILD_SA response 38 [ N(TS_UNACCEPT) ]
````

Below you can see my tries to set up the tunnel. I'm running dozens of tunnels of all kinds on a various network equipment, but I'm totally unsuccessful to set a gre one behind NAT (VTI on the same VM seems to run just fine, but in this case remote party only support gre/ipinip encapsulation). 

````
connections {
    sarov {
        local {
            id = 10.10.1.13
            auth = psk
        }
        remote {
            id = 5.227.67.50
            auth = psk
        }
        children {
            sarov {
                life_time = 1h
                #local_ts = 10.10.1.13[gre],158.160.32.170[gre]
                #remote_ts = 5.227.67.50[gre]
                #local_ts = 158.160.32.170[gre]
                #local_ts = 0.0.0.0/0[gre]
                #remote_ts = 5.227.67.50[gre]
                local_ts = 0.0.0.0/0
                remote_ts = 0.0.0.0/0
                mode = transport
                esp_proposals = aes128ctr-sha256
                start_action = start
                dpd_action = clear
            }
        }
        version = 2
        proposals = aes128-sha256-prfsha256-modp2048
        local_addrs = 10.10.1.13
        remote_addrs = 5.227.67.50
        rekey_time = 1h
        dpd_timeout = 30s
        encap = yes
    }
}
````

**Comment by tobiasbrunner** [comment]
The TS have to match the traffic and the configuration on the other end. So what are the source and destination IPs of the GRE tunnel?

**Comment by Giveback-ru** [comment]
The source IP from my end of the tunnel is 10.10.1.13, the remote 5.227.67.50 will see it as 158.160.32.170 if it were unencrypted gre tunnel, but since it's encrypted and thus NAT cannot change the header value - I guess the remote will still see it as  10.10.1.13, right ?

**Reply by tobiasbrunner** [reply]
Only with tunnel mode would the original IP header be preserved. With transport mode, the addresses in the header will change, so the traffic selectors have to be adapted accordingly. Using `dynamic[gre]` automatically uses the endpoint addresses, but if the peer doesn't have a similar feature, you have to configure the public IP as TS explicitly, or you use tunnel mode, configure the inner IPs as TS, and accept the overhead it brings.
